10up/classifai
# Add a Term Cleanup Feature

### Description of the Change

This PR adds in a new Feature, Term Cleanup, that uses the power of embeddings, to compare all terms against each other and then recommending similar terms that could be merged together if desired. This comparison process can be run entirely in WordPress but once you reach around 500 terms, performance really starts to slow down. Similar to the [Smart 404 Feature](https://github.com/10up/classifai/pull/801), support for Elasticsearch (using ElasticPress) has been added to significantly speed things up.

This Feature requires two services in order to function:

1. A Provider that can generate embeddings (right now we support OpenAI and Azure OpenAI)
2. A data storage engine that can also run vector queries (currently supports doing this in WordPress only or using Elasticsearch powered by ElasticPress)

Once the Feature is setup, you can start the comparison process which will send all terms off to the embeddings Provider to generate vector data. If configured, this data is then sent to Elasticsearch and vector queries are run there. If not configured, this data remains in WordPress only and comparisons are done there using that data.

Once the process is complete, you'll be given the opportunity to see all the results and decide which terms you want to merge together (or which terms you want to skip). This process can be re-run whenever needed (i.e. as new terms are added).

Closes #795 

### How to test the Change

1. Enable the Feature with at least one taxonomy turned on
2. Go to the Term Cleanup page and start the process for whatever taxonomy you want
3. Ensure the process completes and results show
4. Ensure you can merge and skip terms
5. If desired, set up Elasticsearch/ElasticPress and test using that

### Changelog Entry

> Added - New Feature, Term Cleanup, which provides the ability to compare all terms from specific taxonomies and showing recommendations for similar terms that can be merged together

### Credits

Props @dkotter, @iamdharmesh, @berkod 

### Checklist:
